### PR TITLE
Cache handling without input tensors mutation

### PR DESCRIPTION
**Pull request by borisfom**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by titu1994 — DISMISSED** [review]
Thanks !

File: nemo/collections/asr/modules/conformer_encoder.py [hunk 1]
```diff
@@ -536,7 +531,7 @@ def forward_internal(
         if cache_last_channel is not None:
             cache_len = self.streaming_cfg.last_channel_cache_size
             cache_keep_size = max_audio_length - self.streaming_cfg.cache_drop_size
-            cache_last_channel_next = torch.zeros_like(cache_last_channel)
+            # cache_last_channel_next = torch.zeros_like(cache_last_channel)
```
Comment (VahidooX) [inline nemo/collections/asr/modules/conformer_encoder.py:534]: Need to drop this line?

File: nemo/collections/asr/modules/conformer_encoder.py [hunk 2]
```diff
@@ -561,19 +556,32 @@ def forward_internal(
...
             if att_mask is not None:
                 att_mask = att_mask[:, cache_len:]
+            # Convert caches from the tensor to list
+            cache_last_time_next = [None] * self.streaming_cfg.last_time_num
+            cache_last_channel_next = [None] * self.streaming_cfg.last_channel_num
 
         for lth, (drop_prob, layer) in enumerate(zip(self.layer_drop_probs, self.layers)):
             original_signal = audio_signal
+            if cache_last_channel is not None:
+                cache_last_channel_cur = cache_last_channel[layer.self_attn._cache_id]
```
Comment (VahidooX) [inline nemo/collections/asr/modules/conformer_encoder.py:566]: With the latest changes, I think there is no need for cache_id and all these stuff anymore. You may just call each layer and just append the next_cache returned in a list. You may pass something like cache_last_channel[lth] to each layer.

File: nemo/collections/asr/modules/conformer_encoder.py [hunk 3]
```diff
@@ -561,19 +556,32 @@ def forward_internal(
...
-                cache_last_time=cache_last_time,
-                cache_last_channel_next=cache_last_channel_next,
-                cache_last_time_next=cache_last_time_next,
+                cache_last_channel=cache_last_channel_cur,
+                cache_last_time=cache_last_time_cur,
             )
+
+            if cache_last_channel_cur is not None:
+                (audio_signal, cache_last_channel_cur, cache_last_time_cur) = audio_signal
+                cache_last_channel_next[layer.self_attn._cache_id] = cache_last_channel_cur
```
Comment (VahidooX) [inline nemo/collections/asr/modules/conformer_encoder.py:582]: Each next cache can be appended to a list. No need to create a predefined list beforehand.

File: nemo/collections/asr/parts/submodules/causal_convs.py [hunk 4]
```diff
@@ -129,21 +129,23 @@ def __init__(
...
-            if cache_next is not None:
-                cache_keep_size = torch.tensor(x.size(-1) - self.cache_drop_size, dtype=torch.int64, device=x.device)
-                cache_keep_size = torch.clip(cache_keep_size, min=1, max=cache_next.size(-1))
-                cache_next[self._cache_id, :, :, :-cache_keep_size] = cache[self._cache_id, :, :, cache_keep_size:]
-                cache_next[self._cache_id, :, :, -cache_keep_size:] = x[:, :, :cache_keep_size]
-        return new_x
-
-    def forward(self, x, cache=None, cache_next=None):
-        x = self.update_cache(x, cache=cache, cache_next=cache_next)
+            # cache_keep_size = x.size(-1) - self.cache_drop_size, dtype=torch.int64, device=x.device)
```
Comment (VahidooX) [inline nemo/collections/asr/parts/submodules/causal_convs.py:139]: Drop these lines?

File: nemo/collections/asr/parts/submodules/causal_convs.py [hunk 5]
```diff
@@ -129,21 +129,23 @@ def __init__(
...
+                x = x[:, :, : -self.cache_drop_size]
+            cache = torch.cat([cache[:, :, x.size(-1) :], x], dim=-1)
+            # print("cache size: ", cache.size())
+        return new_x, cache
+
+    def forward(self, x, cache=None):
+        x, cache = self.update_cache(x, cache=cache)
         x = super().forward(x)
-        return x
+        return x, cache
```
Comment (VahidooX) [inline nemo/collections/asr/parts/submodules/causal_convs.py:151]: Shouldn't we check the cache and return just one output when cache is None?

File: nemo/collections/asr/parts/submodules/conformer_modules.py [hunk 6]
```diff
@@ -169,31 +161,17 @@ def forward(
...
-            x = self.self_attn(
-                query=x,
-                key=x,
-                value=x,
-                mask=att_mask,
-                pos_emb=pos_emb,
-                cache=cache_last_channel,
-                cache_next=cache_last_channel_next,
-            )
+            x = self.self_attn(query=x, key=x, value=x, mask=att_mask, pos_emb=pos_emb, cache=cache_last_channel,)
```
Comment (VahidooX) [inline nemo/collections/asr/parts/submodules/conformer_modules.py:164]: Drop last comma?

File: nemo/collections/asr/parts/submodules/conformer_modules.py [hunk 7]
```diff
@@ -169,31 +161,17 @@ def forward(
...
-            x = self.self_attn(
-                query=x,
-                key=x,
-                value=x,
-                pad_mask=pad_mask,
-                pos_emb=pos_emb,
-                cache=cache_last_channel,
-                cache_next=cache_last_channel_next,
-            )
+            x = self.self_attn(query=x, key=x, value=x, pad_mask=pad_mask, pos_emb=pos_emb, cache=cache_last_channel,)
```
Comment (VahidooX) [inline nemo/collections/asr/parts/submodules/conformer_modules.py:166]: Another last extra comma.

File: nemo/collections/asr/parts/submodules/conformer_modules.py [hunk 8]
```diff
@@ -368,10 +350,7 @@ def forward(self, x, pad_mask=None, cache=None, cache_next=None):
         if pad_mask is not None:
             x = x.float().masked_fill(pad_mask.unsqueeze(1), 0.0)
 
-        if cache is not None:
-            x = self.depthwise_conv(x, cache=cache, cache_next=cache_next)
-        else:
-            x = self.depthwise_conv(x)
+        x, cache = self.depthwise_conv(x, cache=cache)
```
Comment (VahidooX) [inline nemo/collections/asr/parts/submodules/conformer_modules.py:353]: We need to check if cache is None or not and based on that decide on the number of outputs.

File: nemo/collections/asr/parts/submodules/multi_head_attention.py [hunk 9]
```diff
@@ -142,17 +142,18 @@ def forward(self, query, key, value, mask, pos_emb=None, cache=None, cache_next=
...
+    def update_cache(self, key, value, query, cache):
         if cache is not None:
-            key = value = torch.cat([cache[self._cache_id], key], dim=1)
+            key = value = torch.cat([cache, key], dim=1)
             q_keep_size = query.shape[1] - self.cache_drop_size
-            if cache_next is not None:
-                cache_next[self._cache_id, :, :-q_keep_size, :] = cache[self._cache_id, :, q_keep_size:, :]
-                cache_next[self._cache_id, :, -q_keep_size:, :] = query[:, :q_keep_size, :]
-        return key, value, query
+            if cache is not None:
```
Comment (VahidooX) [inline nemo/collections/asr/parts/submodules/multi_head_attention.py:154]: Cache is already None when we get here.

**Review by VahidooX — CHANGES_REQUESTED** [review]

File: nemo/collections/asr/parts/submodules/multi_head_attention.py [hunk 10]
```diff
@@ -142,17 +141,17 @@ def forward(self, query, key, value, mask, pos_emb=None, cache=None, cache_next=
...
         if cache is not None:
-            key = value = torch.cat([cache[self._cache_id], key], dim=1)
+            key = value = torch.cat([cache, key], dim=1)
             q_keep_size = query.shape[1] - self.cache_drop_size
-            if cache_next is not None:
-                cache_next[self._cache_id, :, :-q_keep_size, :] = cache[self._cache_id, :, q_keep_size:, :]
-                cache_next[self._cache_id, :, -q_keep_size:, :] = query[:, :q_keep_size, :]
-        return key, value, query
+            cache = torch.cat([cache[:, q_keep_size:, :], query[:, :q_keep_size, :]], dim=1)
+        return key, value, query, cache
```
Comment (VahidooX) [inline nemo/collections/asr/parts/submodules/multi_head_attention.py:154]: Need to return single output when cache is none.

**Review by VahidooX — CHANGES_REQUESTED** [review]
Another module has missing single vs double outputs.

File: nemo/collections/asr/parts/submodules/multi_head_attention.py [hunk 11]
```diff
@@ -453,7 +456,11 @@ def forward(self, query, key, value, pad_mask, pos_emb, cache=None, cache_next=N
 
                 out[is_index_global_attn_nonzero] += out_global_to_all
 
-        return self.linear_out(out.reshape(n_batch, -1, self.h * self.d_k)[:, :T])
+        ret = self.linear_out(out.reshape(n_batch, -1, self.h * self.d_k)[:, :T])
+        if cache is None:
```
Comment (VahidooX) [inline nemo/collections/asr/parts/submodules/multi_head_attention.py:460]: Do we need it here as ret looks like to be always single?

**Review by VahidooX — CHANGES_REQUESTED** [review]

File: nemo/collections/asr/parts/submodules/multi_head_attention.py (hunk 10 above)
Comment (borisfom) [reply nemo/collections/asr/parts/submodules/multi_head_attention.py:154]: No, why - update_cache() always returns cache (even if None).

File: nemo/collections/asr/parts/submodules/multi_head_attention.py (hunk 11 above)
Comment (borisfom) [reply nemo/collections/asr/parts/submodules/multi_head_attention.py:460]: Ret is always single, yes - but cache comes from update_cache here, not from ret.

**Review by VahidooX — APPROVED** [review]

**Pull request merged**